### PR TITLE
✨ Support more SAST tools

### PR DESCRIPTION
**Pull request by laurentsimon**
This PR addresses https://github.com/ossf/scorecard/issues/1420, where we came to the conclusion that we want more granularity in the SAST check.
This PR does not contain unit tests yet, because there's a lot to discuss and I want to see if we're all OK with this changes.

**WARNING:  please ignore the comments and the calls to Println()**: I will clean up once we have agreed everything else is fine.

This PR does look for:
- Linter tools run on each PR, and reward 3 points if it's the case
- Code analysis tools run on each PR, and reward 4 points if its the case
- Supply-chain analysis tools run on each PR, and reward 3 points if it's the case. Since scorecard is the only supply-chain tool at the moment and we don't advertise PR support, we give 3 points if scorecard is enabled, even if not run on PRs.

This PR gives no additional points if a tool is declared in a workflow but not run on PRs - with the exception of scorecard above.

closes https://github.com/ossf/scorecard/issues/1380
closes https://github.com/ossf/scorecard/issues/1420
closes https://github.com/ossf/scorecard/issues/1580

### CODE REVIEW

File: checks/packaging.go
Comment (laurentsimon) [inline checks/packaging.go:17]: ignore this file, it's mostly typo changes.

File: clients/githubrepo/workflows.go
Comment (laurentsimon) [inline clients/githubrepo/workflows.go:100]: go-github does not support `check_suite_id` as input https://github.com/google/go-github/blob/178169fc04bc4a05ca48f98709cdd02ab5b556e3/github/actions_workflow_runs.go#L56

So I've simply copied their code from https://github.com/google/go-github/blob/178169fc04bc4a05ca48f98709cdd02ab5b556e3/github/actions_workflow_runs.go#L91 until they add support. We should file an issue for it.

File: clients/workflows.go
Comment (laurentsimon) [inline clients/workflows.go:26]: need discussion: pointer or not?

**Review by naveensrinivasan — COMMENTED** [review]
Great, But we need tests!

**Comment by laurentsimon** [comment]
> Great, But we need tests!

so much to discuss before I add the test. Ill add them at the end. Test-man :p

**Comment by laurentsimon** [comment]
cc @evverx

**Comment by evverx** [comment]
> Linter tools run on each PR, and reward 3 points if it's the case
> Supply-chain analysis tools run on each PR, and reward 3 points if it's the case

I don't think that linters and supply-chain tools (combined) should outweigh SAST tools like LGTM, CodeQL or Coverity so it seems to me that they should be downgraded to 1 or something like that. I mean, systemd for example uses superlinter and it finds issues from time to time but I'm not sure it would be fair to say that in terms of security it's almost as important as LGTM or Coverity. It wouldn't be fair if projects using only LGTM received 4 while projects using only superlinter received 3 either.

I'll try to take a closer look later. Thanks!

**Comment by evverx** [comment]
> This PR gives no additional points if a tool is declared in a workflow but not run on PRs

It's not always feasible to run SAST tools on every PR but if they are run daily or weekly it's still useful and should be rewarded I think. For example it takes CodeQL about 40 minutes to analyze systemd so it doesn't make much sense to clog its GHActions pool on every PR but it's still run daily

**Comment by evverx** [comment]
systemd runs LGTM and superlinter on every PR, CodeQL daily and Coverity Scan daily and with this PR applied its score went from 10 to 6. I'm not sure it's what it should get :-)

**Comment by evverx** [comment]
FWIW The curl score went from 7 to 0 with this PR applied.

Anyway, on the whole I think I like the idea of putting various tools into separate groups and assigning different "scores" to them. Though the way those "scores" are assigned depending on the group, frequency and so on needs tweaking I think. Thanks!

**Comment by azeemshaikh38** [comment]
Will take a closer look at this tomo. Overall LGTM.

**Comment by laurentsimon** [comment]
Thanks for the feedback @evverx I totally agree we need tweak the score; and also that tools that run on commits/schedules should be rewarded. Maybe give 70% of the points if run on commits vs PRs. Feel free to suggest.

**Comment by laurentsimon** [comment]
> systemd runs LGTM and superlinter on every PR, CodeQL daily and Coverity Scan daily and with this PR applied its score went from 10 to 6. I'm not sure it's what it should get :-)

you'd actually get a 10 if you install scorecard as an action. It suffices for one linter and one static analysis code (LGTM in your case) to be run on all PRs + scorecard installed to get 10.

**Comment by evverx** [comment]
> Maybe give 70% of the points if run on commits vs PRs

I think it depends on how frequent those runs are. If projects run SAST tools daily (or a few times a day) I think they should be scored higher than projects with weekly or monthly runs.

> you'd actually get a 10 if you install scorecard as an action.

I get that :-) but it's complicated because on the one hand I think scorecard is helpful but on the other hand I don't want to promote OSSF. Other than that looking at a couple of checks that have already been introduced and issues where new checks are outlined it seems to me that scorecard is moving towards "consumers" instead of "producers" (which is understandable) but I'm not sure it will make sense for "producers" to keep it upstream at some point (time will tell I think). Just to clarify, I have absolutely no problems with this direction but somehow tools catered to "consumers" somehow shift things that should be done by "consumers" to "producers"  (the latest example would be https://github.com/google/oss-fuzz/issues/7146) and generally I wouldn't want to annoy maintainers.

File: clients/githubrepo/workflows.go
Comment (azeemshaikh38) [reply clients/githubrepo/workflows.go:100]: Could you file an issue and add a TODO comment here to point to it?

**Review by azeemshaikh38 — APPROVED** [review]

**Comment by azeemshaikh38** [comment]
LGTM. 2 open issues would be - unit tests and addressing @evverx's comment.

File: clients/checkruns.go
Comment (naveensrinivasan) [reply clients/checkruns.go:23]: Can you please add the reason as to why would like to change to a pointer? 

Every change requested should have a reason as to why it is requested. It helps others understand the thought process and the reasoning. This is applicable across all these comments. 

Pointer sematic has drawbacks we have to be clear on why we need it?  `Go` doesn't encourage pointer semantics. Most of the standard library has value semantics

Comment (azeemshaikh38) [reply clients/checkruns.go:23]: Sure. As part of https://github.com/ossf/scorecard/issues/1242, we are in the process of migrating the current data structure to a more consistent format. The consistent formatting lets us use a convention that implies `nil` == `no data`.

File: clients/githubrepo/workflows.go
Comment (naveensrinivasan) [reply clients/githubrepo/workflows.go:73]: Why pointer semantics `*ListWorkflowRunOptions`? I would recommend it being value sematics.

File: checks/fileparser/github_workflow.go
Comment (naveensrinivasan) [inline checks/fileparser/github_workflow.go:301]: Why have a `func` that never returns error as part of the return?

Comment (naveensrinivasan) [reply checks/fileparser/github_workflow.go:301]: Could we remove the `error`?

One line `go` funcs are discouraged. Is there a specific reason to have this in a separate func?

File: clients/githubrepo/checkruns.go
Comment (naveensrinivasan) [inline clients/githubrepo/checkruns.go:67]: Could we please remove this commented code?

File: clients/githubrepo/workflows.go
Comment (naveensrinivasan) [inline clients/githubrepo/workflows.go:133]: Could we please remove this commented code?

File: clients/repo_client.go
Comment (naveensrinivasan) [inline clients/repo_client.go:38]: Could this not be pointer `*ListWorkflowRunOption`? Makes the API cleaner.

**Review by naveensrinivasan — APPROVED** [review]
Thanks. Few comments.

File: checks/sast.go
Comment (mjpieters) [inline checks/sast.go:102]: Did you mean to list CodeQL twice here?

Comment (laurentsimon) [reply checks/sast.go:102]: nope. Thanks for catching!

**Comment by laurentsimon** [comment]
update: I've not tweaked the logic for score computation as follows:
- linter are run on all PRs. Linters are cheap so this seems acceptable. 1 points is awarded if it's the case
- supply-chain tool are used. We don't check if it's run on all PRs. So long as it's defined in a workflow and/or run on at least one commit, we award 2 points for it
- static analysis. Same as for supply-chain, and we award 5 points. If one tool is run on all PRs, we give an additional 2 points.

Wdut?

**Comment by evverx** [comment]
> supply-chain tool are used. We don't check if it's run on all PRs. So long as it's defined in a workflow and/or run on at least one commit, we award 2 points for it

I think supply chain tools should follow the same rules as static analyzers in the sense that if they can't be run on PRs they should be downgraded to 1. It should help to prompt the developers of those tools to make them compatible with the PR workflow (which is supposed to catch issues as early as possible).

**Comment by laurentsimon** [comment]
> > supply-chain tool are used. We don't check if it's run on all PRs. So long as it's defined in a workflow and/or run on at least one commit, we award 2 points for it
> 
> I think supply chain tools should follow the same rules as static analyzers in the sense that if they can't be run on PRs they should be downgraded to 1. It should help to prompt the developers of those tools to make them compatible with the PR workflow (which is supposed to catch issues as early as possible).

good point. I'll do that then. Thank you for your feedback!

**Comment by laurentsimon** [comment]
thoughts on rate limiting 

https://docs.github.com/en/rest/overview/resources-in-the-rest-api#requests-from-github-actions 
```
When using GITHUB_TOKEN, the rate limit is 1,000 requests per hour per repository. For requests to resources that belong to an enterprise account on GitHub.com, GitHub Enterprise Cloud's rate limit applies, and the limit is 15,000 requests per hour per repository.
```

That's fine since each commit should have a different token.

For PAT, https://docs.github.com/en/developers/apps/building-github-apps/rate-limits-for-github-apps
```
GitHub Apps making server-to-server requests use the installation's minimum rate limit of 5,000 requests per hour. If an application is installed on an organization with more than 20 users, the application receives another 50 requests per hour for each user. Installations that have more than 20 repositories receive another 50 requests per hour for each repository. The maximum rate limit for an installation is 12,500 requests per hour.
```

For a repository like `systemd/systemd`, we'd need 30*6 = ~200 API requests, so let's say 300 for each push. 10-15 commits per hours should be fine.

FYI, we hope to be able to support GITHUB_TOKEN in the future. For cron job, highly unlikely we'll ever be able to run SAST check if we apply the changes proposed here.

@azeemshaikh38 any further thought?

**Comment by evverx** [comment]
> For cron job, highly unlikely we'll ever be able to run SAST check if we apply the changes proposed here.

FWIW Looking at https://metrics.openssf.org/grafana/d/default/metric-dashboard?orgId=1&var-PackageURL=pkg%3Agithub/systemd/systemd it seems it doesn't work even without this PR (assuming the cron job is responsible for updating that dashboard)

**Comment by laurentsimon** [comment]
> > For cron job, highly unlikely we'll ever be able to run SAST check if we apply the changes proposed here.
> 
> FWIW Looking at https://metrics.openssf.org/grafana/d/default/metric-dashboard?orgId=1&var-PackageURL=pkg%3Agithub/systemd/systemd it seems it doesn't work even without this PR (assuming the cron job is responsible for updating that dashboard)

you're correct we currently don't run it. Fyi, the data in the dashboard is out of date.
cc @scovetta @inferno-chromium

**Comment by evverx** [comment]
FWIW @laurentsimon with https://github.com/systemd/systemd/commit/74b781de2545c80b988981aace9e04148cf3dfe3 merged it seems scorecard can no longer detect super-linter. As far as I understand the check isn't aware of the slim image: https://github.com/github/super-linter#slim-image

**Comment by evverx** [comment]
Though looking at
```
    {
      "details": [
        "Info: code-analysis tool 'CodeQL' detected in workflow: .github/workflows/codeql-analysis.yml",
        "Warn: linter tool run on 0 commits out of 15 commits",
        "Warn: supply-chain tool run on 0 commits out of 15 commits",
        "Info: code-analysis tool run on the last 15 commits"
      ],
      "reason": "SAST category of tools are not used%!(EXTRA string=code-analysis, string=supply-chain)",
      "name": "SAST",
```
it seems either the commit based analysis doesn't work as expected or it just doesn't look at PRs there. Anyway I'm not sure why it stopped working.

**Comment by azeemshaikh38** [comment]
> Though looking at
> 
> ```
>     {
>       "details": [
>         "Info: code-analysis tool 'CodeQL' detected in workflow: .github/workflows/codeql-analysis.yml",
>         "Warn: linter tool run on 0 commits out of 15 commits",
>         "Warn: supply-chain tool run on 0 commits out of 15 commits",
>         "Info: code-analysis tool run on the last 15 commits"
>       ],
>       "reason": "SAST category of tools are not used%!(EXTRA string=code-analysis, string=supply-chain)",
>       "name": "SAST",
> ```
> 
> it seems either the commit based analysis doesn't work as expected or it just doesn't look at PRs there. Anyway I'm not sure why it stopped working.


This might be a similar issue to https://github.com/ossf/scorecard/issues/1580#issuecomment-1027946049. We use GitHub Search APIs for detecting SAST tools and GitHub's Search APIs likely unreliable. I tried running Scorecard locally on `systemd` and the results varied. We should probably consider updating the `Search` API in `RepoClient` to use locally available code instead. @laurentsimon fyi.

**Comment by evverx** [comment]
> We use GitHub Search APIs for detecting SAST tools and GitHub's Search APIs likely unreliable.

Could it be that because of that the "code review" check is kind of broken as well?
```
  "repo": {
    "name": "github.com/systemd/systemd",
    "commit": "d74da762a3b1edbb7935c3b3a229db601f734125"
  },
  "scorecard": {
    "version": "4.0.1-86-geb0730a",
    "commit": "eb0730ae797377d24676cdedfa27216f0d649091"
  },
...
    {
      "details": null,
      "score": 4,
      "reason": "GitHub code reviews found for 13 commits out of the last 30 -- score normalized to 4",
      "name": "Code-Review",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/eb0730ae797377d24676cdedfa27216f0d649091/docs/checks.md#code-review",
        "short": "Determines if the project requires code review before pull requests (aka merge requests) are merged."
      }
```

**Comment by azeemshaikh38** [comment]
> re: search API in repoClient @azeemshaikh38 do we need this API at all or can we just let checks use the file listing API https://github.com/ossf/scorecard/blob/main/checks/fileparser/listing.go?

It would be nice to support a generic `Search` API, especially for cases which require searching large repos like oss-fuzz like we do [today](https://github.com/ossf/scorecard/blob/eb0730ae797377d24676cdedfa27216f0d649091/checks/fuzzing.go#L55). Having said that, I'm open to trying out `ListFiles` and `GetFileContent` APIs for solving SAST checks if that gets us more consistency.

**Comment by evverx** [comment]
Also it seems scorecard (with `--verbosity Debug --show-details`) no longer shows useful logs that can be used to somewhat dispute its findings. I think it would be great if they could be brought back given the number of false positives it produces mainly due to the GH API as far as I can tell.

**Comment by azeemshaikh38** [comment]
> Could it be that because of that the "code review" check is kind of broken as well?

Code-Review does not use GitHub search APIs. It looks like #1505 changed the behavior here. Commits before that, give `systemd/systemd` a score of 10 for Code-Review. @laurentsimon could you look into this?

**Comment by evverx** [comment]
> Commits before that, give systemd/systemd a score of 10 for Code-Review

I think commits before that are affected by https://github.com/ossf/scorecard/issues/1260 so it doesn't always work as expected either. Anyway, without "debug" logs it's hard to tell what scorecard does and why.

**Comment by evverx** [comment]
Regarding logs it appears something like that was reported in https://github.com/ossf/scorecard/issues/95 and then it got a bit better and now those logs seem to be gone again.

**Comment by azeemshaikh38** [comment]
> > Commits before that, give systemd/systemd a score of 10 for Code-Review
> 
> I think commits before that are affected by #1260 so it doesn't always work as expected either. Anyway, without "debug" logs it's hard to tell what scorecard does and why.

Ok, I think I know whats going on. Some PRs are repeated for `systemd` commits. Its as if the same PR number gets re-used and latest PR data (including reviews) overwrites old data. Here's an example:

 

```
{
              "oid": "d74da762a3b1edbb7935c3b3a229db601f734125",
              "associatedPullRequests": {
                "nodes": [
                  {
                    "repository": {
                      "owner": {
                        "login": "systemd"
                      },
                      "name": "systemd"
                    },
                    "number": 22506,
                    "title": "gpt-auto: some (primarily cosmetic) fixes to backing block device detection in gpt-auto-generator/sd-device",
                    "mergeCommit": {
                      "oid": "d74da762a3b1edbb7935c3b3a229db601f734125"
                    }
                  }
                ]
              }
            },
{
              "oid": "d5cb053cd93d516f516e0b748271b55f9dfb3a29",
              "associatedPullRequests": {
                "nodes": [
                  {
                    "repository": {
                      "owner": {
                        "login": "systemd"
                      },
                      "name": "systemd"
                    },
                    "number": 22506,
                    "title": "gpt-auto: some (primarily cosmetic) fixes to backing block device detection in gpt-auto-generator/sd-device",
                    "mergeCommit": {
                      "oid": "d74da762a3b1edbb7935c3b3a229db601f734125"
                    }
                  }
                ]
              }
            },
```

2 different commits have the same `associatedPR`, while the PR data itself reflects the latest change. This is why Scorecard finds unreviewed commits, because there is no review data associated with these commits. 

The real question is why is this behavior happening on `systemd` at all? @evverx do you have any clues?

**Comment by evverx** [comment]
My guess would be that that's because that PR was merged so the two commits and the "merge" commit ended up in the repository. If it had been squashed or rebased only that commit would have been mapped to that PR. I need to take a closer look.

**Comment by azeemshaikh38** [comment]
> My guess would be that that's because that PR was merged so the two commits and the "merge" commit ended up in the repository. If it had been squashed or rebased only that commit would have been mapped to that PR. I need to take a closer look.

Ah, I think you are right. https://github.com/systemd/systemd/pull/22506 has 2 commits - [d5cb053cd93d516f516e0b748271b55f9dfb3a29](https://github.com/systemd/systemd/pull/22506/commits/d5cb053cd93d516f516e0b748271b55f9dfb3a29) and [f1ad2c9238f03198a05e36167397d24983ebc527](https://github.com/systemd/systemd/pull/22506/commits/f1ad2c9238f03198a05e36167397d24983ebc527). So it shows up as 3 commits - [d5cb053cd93d516f516e0b748271b55f9dfb3a29](https://github.com/systemd/systemd/pull/22506/commits/d5cb053cd93d516f516e0b748271b55f9dfb3a29), [f1ad2c9238f03198a05e36167397d24983ebc527](https://github.com/systemd/systemd/pull/22506/commits/f1ad2c9238f03198a05e36167397d24983ebc527) and a [merge commit](https://github.com/systemd/systemd/commit/d74da762a3b1edbb7935c3b3a229db601f734125).

This is a very helpful find @evverx, thanks! Let me look into figuring out a fix for this.

**Comment by evverx** [comment]
I think I guessed it because the logs where that PR was mentioned were right in front of me :-) It would be great if they could be kept. Thanks!

**Comment by evverx** [comment]
FWIW looking at the "code review" check it seems it can't cover projects like https://github.com/libbpf/libbpf where patches are reviewed on the kernel mailing list and then get synced with the kernel on GitHub. I think it's a corner case probably but maybe scorecard could look for "Signed-Off", "Reviewed-by" or something like that. It can easily be forged though so I'm not sure what to do about that. Anyway, it has nothing to do with the SAST check so I'll show myself out. Thanks!

**Comment by azeemshaikh38** [comment]
> I think I guessed it because the logs where that PR was mentioned were right in front of me :-) It would be great if they could be kept. Thanks!

Sent #1637 for this. Also added more logging, hopefully that helps.

**Comment by evverx** [comment]
> Regarding logs it appears something like that was reported in https://github.com/ossf/scorecard/issues/95 and then it got a bit better and now those logs seem to be gone again.

Looks like it was an overstatement because the logs weren't actually gone. It's just `--verbosity Debug` no longer works: https://github.com/ossf/scorecard/pull/1637#issuecomment-1039813280. Looking at a few PRs that have been opened recently it's a bit sad that unlike Allstar (against which changes are actually tested and which was the reason the logger was changed in the first place) open source maintainers (or generally people rolling out their own scripts) don't seem to be considered "dowstream consumers" scorecard should care about.

**Comment by laurentsimon** [comment]
> > Regarding logs it appears something like that was reported in #95 and then it got a bit better and now those logs seem to be gone again.
> 
> Looks like it was an overstatement because the logs weren't actually gone. It's just `--verbosity Debug` no longer works: [#1637 (comment)](https://github.com/ossf/scorecard/pull/1637#issuecomment-1039813280). Looking at a few PRs that have been opened recently it's a bit sad that unlike Allstar (against which changes are actually tested and which was the reason the logger was changed in the first place) open source maintainers (or generally people rolling out their own scripts) don't seem to be considered "dowstream consumers" scorecard should care about.

we consider them downstream users. @naveensrinivasan is working on adding unit tests and this one will be caught once he's added them. Thanks for reminding us, though. Please create a tracking issue if there's none specifically for the CLI.

**Comment by evverx** [comment]
> @naveensrinivasan is working on adding unit tests and this one will be caught once he's added them.

FWIW looking at the number of tests that have been added relatively recently I have to say it's impressive and I don't doubt that eventually corner cases like this will be covered as well (and I usually completely agree with @naveensrinivasan  when it comes to testing (including fuzzing).

>  Please create a tracking issue if there's none specifically for the CLI.

@laurentsimon I came here to report a bug in the SAST check and then I updated scorecard and ran into another bug in the Code-Review check and while I was testing that PR I ran into yet another bug/feature related to "debug" mode. I'm not sure I can spend so much time reporting scorecard issues. Anyway, I didn't mean to say that scorecard shouldn't be used in general. My takeaway is that it can be used indirectly via the action or Allstar but it's just not stable enough to rely on outside of the OSSF ecosystem (and there is nothing wrong with that). I think I can wait until it's more or less stabilized.

**Comment by azeemshaikh38** [comment]
> Looks like it was an overstatement because the logs weren't actually gone. It's just `--verbosity Debug` no longer works: [#1637 (comment)](https://github.com/ossf/scorecard/pull/1637#issuecomment-1039813280). Looking at a few PRs that have been opened recently it's a bit sad that unlike Allstar (against which changes are actually tested and which was the reason the logger was changed in the first place) open source maintainers (or generally people rolling out their own scripts) don't seem to be considered "dowstream consumers" scorecard should care about.

I hear you @evverx. It's an unfortunate result of us not having tests like Laurent mentioned. #1650 should fix the problem you are having. Let us know if we can help in any other way.

File: clients/workflows.go
Comment (laurentsimon) [reply clients/workflows.go:19]: will keep as-is because the github-go API use a string. I don't disagree with the suggestion though

File: clients/githubrepo/workflows.go
Comment (laurentsimon) [reply clients/githubrepo/workflows.go:155]: I made it a pointer. Pointers are better for large structure, right? Otherwise it's copying around data for calls, etc, I think

File: clients/checkruns.go
Comment (laurentsimon) [reply clients/checkruns.go:23]: If I make `Name` a pointer, shall I make everything else one too then?

**Comment by laurentsimon** [comment]
@azeemshaikh38 take a final look before I merge. Ignore the comments, I'm going to remove them before the merge.

**Comment by evverx** [comment]
FWIW Looking at https://github.com/ossf/scorecard/pull/1816 I think by that logic the SAST check should be inconclusive as well when scorecard can't find static analyzers it's aware of. Coverity is often hidden in bash scripts sending data to the scanner and it's unlikely that `scorecard` can ever detect that reliably.

**Comment by azeemshaikh38** [comment]
@laurentsimon are we blocked on anything to get this merged?

**Comment by laurentsimon** [comment]
no real blocker, mostly delayed to fix the e2e tests.

**Comment by laurentsimon** [comment]
Please don't close, Im going to get to this PR sometimes after I've migrated other checks to raw results.

**Comment by naveensrinivasan** [comment]
> Please don't close, Im going to get to this PR sometimes after I've migrated other checks to raw results.

I apologize it was a mistake.

**Comment by laurentsimon** [comment]
> > Please don't close, Im going to get to this PR sometimes after I've migrated other checks to raw results.
> 
> I apologize it was a mistake.

np, happens to be all the time :)

**Comment by naveensrinivasan** [comment]
@laurentsimon  What do we want to do about this PR?

**Comment by laurentsimon** [comment]
Let's keep it open if you don't mind. The code uses too many API calls, but I think if we relax some of checks, a large part of the code can be re-used. I'm not working on it atm, though.

**Comment by naveensrinivasan** [comment]
> Let's keep it open if you don't mind. The code uses too many API calls, but I think if we relax some of checks, a large part of the code can be re-used. I'm not working on it atm, though.

OK, sounds good! Thanks

**Comment by naveensrinivasan** [comment]
Hi @laurentsimon,

 I wondered if keeping this Pull Request open would be beneficial, considering it is over a year old.

**Comment by naveensrinivasan** [comment]
> Hi @laurentsimon,
> 
> I wondered if keeping this Pull Request open would be beneficial, considering it is over a year old.

A friendly ping @laurentsimon

**Comment by laurentsimon** [comment]
> > Hi @laurentsimon,
> > I wondered if keeping this Pull Request open would be beneficial, considering it is over a year old.
> 
> A friendly ping @laurentsimon

Up to you. There's a lot of useful code someone may be able to re-use to implement the SAST. But we can close it a search for it if need be.

**Comment by naveensrinivasan** [comment]
> > > Hi @laurentsimon,
> > > I wondered if keeping this Pull Request open would be beneficial, considering it is over a year old.
> > 
> > 
> > A friendly ping @laurentsimon
> 
> Up to you. There's a lot of useful code someone may be able to re-use to implement the SAST. But we can close it a search for it if need be.

Thanks for keeping it clean. I will close it.

**Pull request closed without merging**
